src/storage/txn/scheduler.rs

```diff
@@ -11,92 +11,593 @@
+    schedch: SchedCh,
+
+    // cmd id generator
+    idalloc: u64,
+
+    // simulate memory row locks
+    rowlocks: MemRowLocks,
+
+    // statistics for flow control
+    running_cmd_count: u64,
+}
+
+fn readonly_cmd(cmd: &Command) -> bool {
+    match *cmd {
+        Command::Get { .. } |
+        Command::BatchGet { .. } |
+        Command::Scan { .. } => true,
+        _ => false,
+    }
 }
 
 impl Scheduler {
-    pub fn new(engine: Arc<Box<Engine>>) -> Scheduler {
+    pub fn new(engine: Arc<Box<Engine>>, schedch: SchedCh, concurrency: usize) -> Scheduler {
         Scheduler {
-            store: Arc::new(TxnStore::new(engine)),
-            pool: ThreadPool::new_with_name(thd_name!("txn-scheduler-pool"), DEFAULT_POOL_SIZE),
+            engine: engine,
+            cmd_ctxs: HashMap::new(),
+            schedch: schedch,
+            idalloc: 0,
+            rowlocks: MemRowLocks::new(concurrency),
+            running_cmd_count: 0,
         }
     }
 
-    pub fn exec(&self, cmd: Command) {
-        let store = self.store.clone();
-        self.pool.execute(move || handle_cmd(store, cmd));
+    fn next_id(&mut self) -> u64 {
+        self.idalloc += 1;
+        self.idalloc
     }
-}
 
-fn handle_cmd(store: Arc<TxnStore>, cmd: Command) {
-    let cmd_str = format!("{}", cmd);
-    debug!("scheduler::handle_cmd begin: {}", cmd_str);
-    match cmd {
-        Command::Get { ctx, key, start_ts, callback } => {
-            callback(store.get(ctx, &key, start_ts).map_err(::storage::Error::from));
-        }
-        Command::BatchGet { ctx, keys, start_ts, callback } => {
-            callback(match store.batch_get(ctx, &keys, start_ts) {
-                Ok(results) => {
-                    let mut res = vec![];
-                    for (k, v) in keys.into_iter().zip(results.into_iter()) {
-                        match v {
-                            Ok(Some(x)) => res.push(Ok((k.raw().unwrap(), x))),
-                            Ok(None) => {}
-                            Err(e) => res.push(Err(::storage::Error::from(e))),
-                        }
+    fn incr_running_cmd(&mut self) {
+        self.running_cmd_count += 1;
+    }
+
+    fn decr_running_cmd(&mut self) {
+        self.running_cmd_count -= 1;
+    }
+
+    fn save_cmd_context(&mut self, cid: u64, ctx: RunningCtx) {
+        if let Some(_) = self.cmd_ctxs.insert(cid, ctx) {
+            panic!("cid = {} existed, fatal", cid);
+        }
+    }
+
+    fn finish_cmd(&mut self, cid: u64) {
+        {
+            // running ctx must existed
+            let rctx = self.cmd_ctxs.get(&cid).unwrap();
+            assert_eq!(cid, rctx.cid);
+
+            // release lock and wake up waiting commands
+            if !rctx.needed_locks().is_empty() {
+                let wakeup_list = self.rowlocks.release_by_indexs(rctx.needed_locks(), cid);
+                for wcid in wakeup_list {
+                    if let Err(e) = self.schedch.send(Msg::AcquireLock { cid: wcid }) {
```

Comment:
I think it's better to keep control flow clear, call acquire_lock_for_cmd(wcid) is not finish_cmd's job.